File: 02/one.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func checksum(ids []string) int {
	twos := 0
	threes := 0

	for _, id := range ids {
		freq := map[rune]int{}
		for _, ch := range id {
			freq[ch]++
		}

		hasTwo := false
		hasThree := false
		for _, count := range freq {
			if count == 2 {
				hasTwo = true
			}
			if count == 3 {
				hasThree = true
			}
		}

		if hasTwo {
			twos++
		}
		if hasThree {
			threes++
		}
	}

	return twos * threes
}

func main() {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	ids := strings.Split(strings.TrimSpace(string(data)), "\n")
	fmt.Println(checksum(ids))
}

File: 02/one.hs
import Data.List (group, sort)

charFreqs :: String -> [Int]
charFreqs = map length . group . sort

hasExactly :: Int -> String -> Bool
hasExactly n = elem n . charFreqs

checksum :: [String] -> Int
checksum ids = length (filter (hasExactly 2) ids) * length (filter (hasExactly 3) ids)

main :: IO ()
main = do
    input <- readFile "input.txt"
    let ids = lines input
    print (checksum ids)
